KEYCLOAK-6630 Client scopes initial support

mposolda opened the pull request.

mposolda committed:
KEYCLOAK-6630 KEYCLOAK-349 Client Scopes

Co-authored-by: vramik <vramik@redhat.com>

mposolda commented:
I've added some tests to PR. It's ready for review now.

vramik commented:
in client-scope-mappings.html there is `<kc-tooltip>Client template has full scope allowed, which means this client will have the full scope of all roles.</kc-tooltip>` I'm not sure if it is actually used somewhere in admin console (I wasn't able to find it yet)

vramik commented:
on Oracle I get: Error : 1408, Position : 70, Sql = CREATE INDEX DBALLO01.IDX_CLSCOPE_PROTMAP ON DBALLO01.PROTOCOL_MAPPER(CLIENT_SCOPE_ID), OriginalSql = CREATE INDEX DBALLO01.IDX_CLSCOPE_PROTMAP ON DBALLO01.PROTOCOL_MAPPER(CLIENT_SCOPE_ID), Error Msg = ORA-01408: such column list already indexed

vramik commented:
on MSSQL I get ERROR [org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider] (ServerService Thread Pool -- 53) Change Set META-INF/jpa-changelog-4.0.0.xml::4.0.0-KEYCLOAK-5579::mposolda@redhat.com failed.  Error: The object 'DF_CLIENT_TEMPLATE_FULL_SCOPE_ALLOWED' is dependent on column 'FULL_SCOPE_ALLOWED'. [Failed SQL: ALTER TABLE [dbo].[CLIENT_TEMPLATE] DROP COLUMN [FULL_SCOPE_ALLOWED]]: liquibase.exception.DatabaseException: The object 'DF_CLIENT_TEMPLATE_FULL_SCOPE_ALLOWED' is dependent on column 'FULL_SCOPE_ALLOWED'. [Failed SQL: ALTER TABLE [dbo].[CLIENT_TEMPLATE] DROP COLUMN [FULL_SCOPE_ALLOWED]]
...
...
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: The object 'DF_CLIENT_TEMPLATE_FULL_SCOPE_ALLOWED' is dependent on column 'FULL_SCOPE_ALLOWED'.

mposolda commented:
@vramik Thanks for the review and DB testing! Agree with everything you pointed. Regarding client-scope-mappings.html, it looks that whole form "allowScope" (line 12-32 of file client-scope-mappings.html) can be removed. This page is shown when you are on detail of some client-scope in admin console and you click on tab "Scope" (the one with Role Scope Mappings).

vramik commented:
@mposolda Regarding to Oracle, there is the [index](https://github.com/keycloak/keycloak/blob/fe98c30077d49822624c728e9fe668f8fb0ff074/model/jpa/src/main/resources/META-INF/jpa-changelog-3.2.0.xml#L152-L154) what it complains about

mposolda commented:
Failing travis cross-dc test is due the https://issues.jboss.org/browse/KEYCLOAK-7017 . It's not related to this PR.

mposolda commented:
Documentation PR: https://github.com/keycloak/keycloak-documentation/pull/389

mposolda commented:
@fabianschlier yes, you're right. The client scopes reworks this topic and it will allow to limit used protocolMappers based on the scope parameter. Not sure when this will be available in keycloak master. Hopefully very soon (in the next release).

mposolda commented:
@vramik I've did the rebase with latest master and needed to squash to single commit. Could you please re-run the DB matrix? You're welcome to re-review the PR again after rebase if you have time. But there are not much changes, mostly just the things related to the rebase.

vramik commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java:
```diff
@@ -1287,13 +1287,13 @@ public boolean removeGroup(GroupModel group) {
-        List<String> clientTemplates = cached.getClientTemplates();
+    public List<ClientScopeModel> getClientScopes() {
+        if (isUpdated()) return updated.getClientScopes();
+        List<String> clientTemplates = cached.getClientScopes();
```
clientScopes

vramik commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java:
```diff
@@ -1287,13 +1287,13 @@ public boolean removeGroup(GroupModel group) {
-            ClientTemplateModel model = cacheSession.getClientTemplateById(id, this);
+            ClientScopeModel model = cacheSession.getClientScopeById(id, this);
             if (model == null) {
                 throw new IllegalStateException("Cached clientemplate not found: " + id);
```
clientScopes

vramik commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/stream/HasRolePredicate.java:
```diff
@@ -55,8 +55,8 @@ public boolean test(Map.Entry<String, Revisioned> entry) {
-        if (value instanceof CachedClientTemplate) {
-            CachedClientTemplate cachedClientTemplate = (CachedClientTemplate)value;
+        if (value instanceof CachedClientScope) {
+            CachedClientScope cachedClientTemplate = (CachedClientScope)value;
```
cachedClientTemplate

vramik commented on server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo2_3_0.java:
```diff
@@ -44,7 +44,7 @@ protected void migrateRealm(RealmModel realm) {
         }
 
-        for (ClientTemplateModel clientTemplate : realm.getClientTemplates()) {
+        for (ClientScopeModel clientTemplate : realm.getClientScopes()) {
```
clientScope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java:
```diff
@@ -66,13 +66,13 @@
     void requireView(ClientModel client);
 
-    boolean canManage(ClientTemplateModel template);
+    boolean canManage(ClientScopeModel template);
```
scope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java:
```diff
@@ -66,13 +66,13 @@
+    boolean canManage(ClientScopeModel template);
 
-    void requireManage(ClientTemplateModel template);
+    void requireManage(ClientScopeModel template);
```
scope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java:
```diff
@@ -66,13 +66,13 @@
+    void requireManage(ClientScopeModel template);
 
-    boolean canView(ClientTemplateModel template);
+    boolean canView(ClientScopeModel template);
```
scope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java:
```diff
@@ -66,13 +66,13 @@
+    boolean canView(ClientScopeModel template);
 
-    void requireView(ClientTemplateModel template);
+    void requireView(ClientScopeModel template);
```
scope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java:
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
 
     @Override
-    public boolean canManage(ClientTemplateModel template) {
+    public boolean canManage(ClientScopeModel template) {
```
scope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java:
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
 
     @Override
-    public void requireManage(ClientTemplateModel template) {
+    public void requireManage(ClientScopeModel template) {
```
scope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java:
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
 
     @Override
-    public boolean canView(ClientTemplateModel template) {
+    public boolean canView(ClientScopeModel template) {
```
scope

vramik commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java:
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
 
     @Override
-    public void requireView(ClientTemplateModel template) {
+    public void requireView(ClientScopeModel template) {
```
scope

vramik commented on testsuite/utils/src/main/java/org/keycloak/testsuite/util/cli/TestCacheUtils.java:
```diff
@@ -48,8 +48,8 @@ public static void cacheRealmWithEverything(KeycloakSession session, String real
 
-        for (ClientTemplateModel clientTemplate : realm.getClientTemplates()) {
-            realm.getClientTemplateById(clientTemplate.getId());
+        for (ClientScopeModel clientTemplate : realm.getClientScopes()) {
```
clientScope

vramik commented:
@mposolda DB matrix passed

mposolda committed:
KEYCLOAK-6630 Change some clientTemplate occurences to clientScope

mposolda commented:
@vramik Thanks, I've updated based on your feedback. I've changed all the methods on ClientPermissionEvaluator interface to reflect "clientScopes" instead of "templates" . Keep the changes in separate commit for easier review.

frekele commented:
Hi @mposolda ,

after upgrading keycloak 3.4.3.Final to v4.0.0-Final and soon after v4.1.0-Final, it is not possible to use Keycloak admin cli as in the example:
https://www.keycloak.org/docs/latest/server_development/index.html#example-using-java

```java
Keycloak keycloak = Keycloak.getInstance(
    "http://localhost:8080/auth",
    "master",
    "admin",
    "password",
    "admin-cli");
RealmRepresentation realm = keycloak.realm("master").toRepresentation();
```

Cause a NullPointerException in [BearerAuthFilter.java line 53](https://github.com/keycloak/keycloak/blob/4.1.0.Final/integration/admin-client/src/main/java/org/keycloak/admin/client/resource/BearerAuthFilter.java#L53)
```java
Caused by: java.lang.NullPointerException
	at org.keycloak.admin.client.resource.BearerAuthFilter.filter(BearerAuthFilter.java:53)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:573)
	... 45 more
```
authHeader is null because tokenService.grantToken call return all fields as empty in [TokenManager.java#L89](https://github.com/keycloak/keycloak/blob/4.1.0.Final/integration/admin-client/src/main/java/org/keycloak/admin/client/token/TokenManager.java#L89)

![currenttoken](https://user-images.githubusercontent.com/2288215/42393855-58c85e50-812e-11e8-9b7e-7dc9d532d8e0.png)

In keycloak 3.4.3.Final it works perfectly.


I do not know if this PR solve this problem, so I'm adding here my comment, did not want to open an issue because maybe this PR already solves this?!?!

Thank you
